### DISCUSSION TITLE
How to test next-auth middleware?

### DISCUSSION THREAD

**Discussion by flacial** [garageScript/c0d3-app#1600] [Q&A] [answered]
I'm having difficulties testing its middleware. Usually when I write the tests of something, it has functions I can mock and test if they were just called with our without certain value. In the case of next-auth, I'd like to test the signIn callback, but:

`NextAuth` returns this:

```js
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 121983,
  [Symbol(trigger_async_id_symbol)]: 121973,
  [Symbol(destroyed)]: { destroyed: false }
}
```

Oh, it's a promise, it makes sense to `await` or `.then` it:

```js
undefined
```

Even its function return type is `any`. So, I'm not certain. Is there a way to test or we should ignore it in the code coverage somehow?

Useful link of it being tested with Cypress: https://next-auth.js.org/tutorials/testing-with-cypress I think Cypress is used for E2E tests which [we don't have](https://github.com/garageScript/c0d3-app/pull/992) yet I think.

**Comment by songz** [comment, marked as answer]
FYI When struggling with `unit` test, the correct solution isn't "let's add functional test". It works, but it hides the underlying problem which is that we want to figure out a way to unit test the function.

I don't think we need to test the return value of nextauth middleware, but rather what we pass into nextauth right via [the callbacks](https://next-auth.js.org/configuration/callbacks)? If so, we just need to test the callback functions.

Example (about signInCallback)

```js
export const signInCallback = async ({ user, account, profile, email, credentials }) {
      ...
}

...
  callbacks: {
    signIn: signinCallback,
    ...
...
}
```

In your test you'd just test `signInCallback` right?

**Reply by flacial** [reply]
Yeah. Thanks for the help!
